ValidatorSetSig Contract (#404)
ava-labs/teleporter: contracts/src/Utilities/README.md

Reviewer: Can you include a description of how the off-chain Warp message is expected to be formatted? I believe the steps are:
1.  ABI-encode a `ValidatorSetSigMessage`
2. Pack that as the payload of an `[AddressedCall](https://github.com/ava-labs/avalanchego/blob/0c4efd743e1d737f4e8970d0e0ebf229ea44406c/vms/platformvm/warp/payload/addressed_call.go#L15)` Warp Message format
3. Pack the `AddressedCall` as the payload  of the `[UnsignedWarpMessage](https://github.com/ava-labs/avalanchego/blob/f17ea6a7ab4036c41b693e47b94d8f0c81cb69ec/vms/platformvm/warp/unsigned_message.go#L14)`

## Utility Contracts

### ValidatorSetSig Contract

This contract provides an alternative to MultiSig contracts for upgradeability where instead of requiring k of n signatures on an upgrade transaction we can require a quorum of validators from a given blockchain.

> The contract accepts off-chain [Avalanche Warp Messages](https://docs.avax.network/build/cross-chain/awm/overview) signed via a manual approve by a chain's validators. It requires these messages to have source address set to the zero address to enforce this. 

Reviewer: ```suggestion
The contract accepts off-chain [Avalanche Warp Messages](https://docs.avax.network/build/cross-chain/awm/overview), which are manually approved for signing  by a chain's validators. It requires these messages to have the source address set to the zero address to enforce this. 
```

Note:
1. The blockchain validating the message may or may not be the same chain where the target contract and the ValidatorSetSig contract are deployed.

> 2. The quorum of signers required for a Warp Message to be considered valid is configured by the validators themselves via their [Warp configs](https://github.com/ava-labs/subnet-evm/issues/729)

Reviewer: This links to the issue describing off-chain Warp messages, which is distinct from a chain's stake weight threshold requirement for signature verification. I don't think we need to mention signature verification here, as it's covered extensively elsewhere, but keeping the link and rewording the bullet point to describe off-chain Warp messages in detail would be good to include.

> ## Utility Contracts

Reviewer 2: nit: thoughts on changing directoyr name `Utilities` -> `utils` or `utilities`? My pref is `utils` but might look off for the abi-binding generated file

Author: Right now I prefer `Utilities` to match `Mocks` and `Teleporter` directories as we have them now. We do have #410 issue open and if we want to lower-case all of them we can do it there?

Reviewer 2: sgtm

### ValidatorSetSig Contract

> This contract provides an alternative to traditional multi-signature contracts where instead of requiring signatures from `K` of `N` pre-specified signers, an aggregate signature is required from a quorum of the current validators a given blockchain.

Reviewer 2: ```suggestion
This contract provides an alternative to traditional multi-signature contracts where instead of requiring signatures from `K` of `N` pre-specified signers, an aggregate signature is required from a quorum of the current validators for a given blockchain.
```

> The contract leverages off-chain [Avalanche Warp Messages](https://docs.avax.network/build/cross-chain/awm/overview), which are manually approved for signing  by a chain's validators. It requires these messages to have the source address set to the zero address to enforce this.

Reviewer 2: ```suggestion
The contract leverages off-chain [Avalanche Warp Messages](https://docs.avax.network/build/cross-chain/awm/overview), which are manually approved for signing by a chain's validators. It requires these messages to have the source address set to the zero address to enforce this.
```

Note:

> 1. The blockchain validating the message may or may not be the same chain where the target contract and the `ValidatorSetSig` contract are deployed.

Reviewer 3: ```suggestion
1. The blockchain validating the message may or may not be the same chain where the target contract and the `ValidatorSetSig` contract is deployed.
```

Author: Both target contract and the validator  are included in the sentence above so `are` is correct.

Reviewer 3: Oh yeah, my bad 😓

2. [Off-Chain Warp messages](https://github.com/ava-labs/subnet-evm/issues/729) are described in the linked issue. They are Warp messages that validators can include in their config to indicate that they are willing to sign them even though they are not a result of on-chain activity.

> #### Creating a valid Off-chain Warp Message for interaction with the ValidatorSetSig Contract

Reviewer 2: nit: should we keep capitalization for "Off-chain" same throughout?

1. ABI-encode a `ValidatorSetSigMessage` as defined in `ValidatorSetSig.sol`
2. OPTIONAL: call `validateMessage` view method of the intended contract to confirm that the message has correct fields, including the nonce.
3. Pack that as the payload of na [AddressedCall](https://github.com/ava-labs/avalanchego/blob/0c4efd743e1d737f4e8970d0e0ebf229ea44406c/vms/platformvm/warp/payload/addressed_call.go#L15) Warp Message format.

> 4. Pack the `AddressedCall` as the payload of the [UnsignedWarpMessage](https://github.com/ava-labs/avalanchego/blob/f17ea6a7ab4036c41b693e47b94d8f0c81cb69ec/vms/platformvm/warp/unsigned_message.go#L14).

Reviewer 2: should we reference aggregating signatures, and possibly point to awm-relayer here

Reviewer: I don't think that's necessary since signature aggregation is separate from the validator signing the message provided via config

Reviewer 2: I'm asking about including signature aggregation because the section `Creating a valid Off-chain Warp Message for interaction with the ValidatorSetSig Contract` is for "valid off-chain Warp message", and semantically we've used "valid Warp message" as one that includes valid aggregate signature

Reviewer 2: would suggest removing "valid" from the header, but no strong pref

Reviewer: Fair point, though I think the word "valid" should still be included. If we want to be technically accurate, what we're doing is providing a "valid unsigned off-chain Warp message". I don't think it's necessary to mention signature aggregation or any of the other details around actual message delivery here. Or if we do, keep it high level and say that signatures can be aggregated and packaged into a transaction in the same manner as any other Warp message.

Reviewer 2: i agree high level is good enough

### ValidatorSetSig Contract

This contract provides an alternative to traditional multi-signature contracts where instead of requiring signatures from `K` of `N` pre-specified signers, an aggregate signature is required from a quorum of the current validators a given blockchain.

The contract leverages off-chain [Avalanche Warp Messages](https://docs.avax.network/build/cross-chain/awm/overview), which are manually approved for signing  by a chain's validators. It requires these messages to have the source address set to the zero address to enforce this.

Note:
1. The blockchain validating the message may or may not be the same chain where the target contract and the `ValidatorSetSig` contract are deployed.

> 2. [Off-Chain Warp messages](https://github.com/ava-labs/subnet-evm/issues/729) are described in the linked issue. They are Warp messages that validators can include in their config to indicate that they are willing to sign them even though they are not a result of on-chain activity.

Reviewer: ```suggestion
2. [Off-Chain Warp messages](https://github.com/ava-labs/subnet-evm/issues/729) are Warp messages that validators can include in their config to indicate that they are willing to sign them even though they are not a result of on-chain activity.
```

#### Creating a valid Off-chain Warp Message for interaction with the ValidatorSetSig Contract

1. ABI-encode a `ValidatorSetSigMessage` as defined in `ValidatorSetSig.sol`
2. OPTIONAL: call `validateMessage` view method of the intended contract to confirm that the message has correct fields, including the nonce.

> 3. Pack that as the payload of na [AddressedCall](https://github.com/ava-labs/avalanchego/blob/0c4efd743e1d737f4e8970d0e0ebf229ea44406c/vms/platformvm/warp/payload/addressed_call.go#L15) Warp Message format.

Reviewer: ```suggestion
3. Pack that as the payload of an [AddressedCall](https://github.com/ava-labs/avalanchego/blob/0c4efd743e1d737f4e8970d0e0ebf229ea44406c/vms/platformvm/warp/payload/addressed_call.go#L15) Warp Message format.
```
